drivers/misc/morse_code/morse_code.c

```diff
@@ -0,0 +1,406 @@
+	const struct morse_code_config *const cfg = mdev->config;
+	struct morse_code_data *ctx = mdev->data;
+	int bit_state;
+
+	if (ctx->data_idx == ctx->data_size) {
+		LOG_DBG("Finish transmission");
+		counter_stop(dev);
+		gpio_pin_configure_dt(&cfg->led, GPIO_OUTPUT_INACTIVE);
+		ctx->cb_info.status = 0;
+		k_work_submit(&ctx->cb_work);
+		return;
+	}
+
+	if (ctx->code_fsm == MORSE_CODE_STREAM_FSM_LOAD) {
+		if (morse_code_load(ctx)) {
+			counter_stop(dev);
+			ctx->cb_info.status = -EINVAL;
+			k_work_submit(&ctx->cb_work);
+			return;
+		}
+	}
+
+	bit_state = morse_code_peak_bit_state(ctx);
+	gpio_pin_configure_dt(&cfg->led, bit_state ? GPIO_OUTPUT_ACTIVE
+						   : GPIO_OUTPUT_INACTIVE);
+	LOG_DBG("%d", bit_state);
+}
+
+int morse_code_send(const struct device *dev, const uint8_t *data, const size_t size)
+{
+	if (dev == NULL || (data == NULL && size != 0)) {
+		LOG_ERR("Device or Data is invalid");
+		return -EINVAL;
+	}
+
+	const struct morse_code_config *const cfg = dev->config;
+	struct morse_code_data *ctx = dev->data;
+
+	if (size == 0 || ctx->data_idx != ctx->data_size) {
+		return (ctx->data_idx != ctx->data_size) ? -EBUSY : 0;
+	}
+
+	LOG_HEXDUMP_DBG(data, size, "data:");
+
+	ctx->data = data;
+	ctx->data_size = size;
+	ctx->data_idx = 0;
+
+	ctx->code_fsm = MORSE_CODE_STREAM_FSM_LOAD;
+	ctx->bit_fsm = MORSE_CODE_BIT_FSM_LOAD_SYMBOL;
+	if (morse_code_load(ctx)) {
+		LOG_ERR("No morse code entry");
+		return -ENOENT;
+	}
+
+	counter_start(cfg->timer);
+
+	return 0;
+}
+
+int morse_code_manage_callback(const struct device *dev, morse_callback_handler_t *cb, void *ctx)
+{
+	if (dev == NULL) {
+		return -EINVAL;
+	}
+
+	struct morse_code_data *drv_ctx = dev->data;
+
+	drv_ctx->cb_info.callback = cb;
+	drv_ctx->cb_info.ctx = cb ? ctx : NULL;
+	drv_ctx->cb_info.status = 0;
+
+	return 0;
+}
+
+int morse_code_set_config(const struct device *dev, uint16_t speed)
+{
+	const struct morse_code_config *cfg;
+	struct morse_code_data *ctx;
+	uint32_t dot_time;
+	int ret;
+
+	ret = morse_code_send(dev, NULL, 0);
+	if (ret) {
+		return ret;
+	}
+
+	cfg = dev->config;
+	ctx = dev->data;
+
+	/* 60s/50wps * speed */
+	dot_time = 60000000U / (50U * speed);
+
+	ctx->dot_tick.ticks = counter_us_to_ticks(cfg->timer, dot_time);
+
+	ret = counter_set_top_value(cfg->timer, &ctx->dot_tick);
+	if (ret) {
+		LOG_ERR("Error at counter_set_top_value %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Device %s ready. Tick: %d", dev->name, dot_time);
+
+	return 0;
+}
+
+static int morse_code_init(const struct device *dev)
+{
+	const struct morse_code_config *const cfg = dev->config;
+	struct morse_code_data *ctx = dev->data;
+	int ret;
+
+	LOG_DBG("Timer");
+	if (!device_is_ready(cfg->timer)) {
+		LOG_ERR("Error: Timer device %s is not ready", cfg->timer->name);
+		return -ENODEV;
+	}
+
+	LOG_DBG("LED");
+	if (!gpio_is_ready_dt(&cfg->led)) {
+		LOG_ERR("Error: LED device %s is not ready", cfg->led.port->name);
+		return -ENODEV;
+	}
```

Comment:
Again, rework for generic I/O, instead of LEDs specifically?

Reply:
dito